### PR TITLE
Add retry functionality

### PR DESCRIPTION
**Pull request by caspervdw**
This adds functionality for retrying requests to the threedi-api-client. Example usage:

```
from threedi_api_client import ThreediApi

api = ThreediApi(
    env_file=".env",
    version="v3-beta",
    rest_config={"retries": 3},  # or a urllib3.util.Retry instance for more functionality
)
```

For asynchronous operation, I had to hack a bit into the generated client. I made use of aiohttp_retry. To minimize the access to private properties (which would give instability across versions) I integrated the code into threedi-api-client and adapted the call signature of the `RetryClient` so that it wraps the `ClientSession` which is initialized by the generated code. This now works:

```
from threedi_api_client import ThreediApi
from threedi_api_client.aio.aiohttp_retry import ExponentialRetry

api = ThreediApi(
    env_file=".env",
    version="v3-beta",
    rest_config={"retries": ExponentialRetry(attempts=3)}, 
)
```

### CODE REVIEW

File: threedi_api_client/aio/aiohttp_retry.py [hunk 1]
```diff
@@ -0,0 +1,390 @@
...
+    async def __aenter__(self) -> ClientResponse:
+        return await self._do_request()
+
+    async def __aexit__(self, exc_type: Any, exc_val: Any, exc_tb: Any) -> None:
+        if self._response is not None:
+            if not self._response.closed:
+                self._response.close()
+
+
+def _url_to_urls(url: _URL_TYPE, attempts: int) -> Tuple[StrOrURL, ...]:
```
Comment (jpprins1) [inline threedi_api_client/aio/aiohttp_retry.py:198]: Why do you need to create a list of urls from a url?

**Review by jpprins1 — APPROVED** [review]
Looks good, just have some questions (for my understanding):
- Why do you need to create a list of urls from a single url?
- Did you override the whole rest stuff to provide a more granular interface for setting the retry policy? 
- Subclassing the `RestClientObject` won't work? ( https://github.com/nens/threedi-api-client/blob/b49b7481e7b83d2d0fde64153559ce3e31085ede/threedi_api_client/openapi/rest.py#L51)

**Comment by jpprins1** [comment]
What's your opinion about using one of the retry policies as default?

File: threedi_api_client/aio/aiohttp_retry.py (hunk 1 above)
Comment (caspervdw) [reply threedi_api_client/aio/aiohttp_retry.py:198]: No idea, this is just a copy of aiohttp_retry

**Comment by caspervdw** [comment]
> Looks good, just have some questions (for my understanding):
> 
>     * Why do you need to create a list of urls from a single url?

This is part of the code from aiohttp_retry, you probably had a closer look then I have had.

> 
>     * Did you override the whole rest stuff to provide a more granular interface for setting the retry policy?

For the async client, the `configuration.retries` was just ignored. I had to override the `pool_manager` so that something happened.

>     * Subclassing the `RestClientObject` won't work? ( https://github.com/nens/threedi-api-client/blob/b49b7481e7b83d2d0fde64153559ce3e31085ede/threedi_api_client/openapi/rest.py#L51
>       )

I don't know how to inject the subclass then, this is also in generated code:
https://github.com/nens/threedi-api-client/blob/b49b7481e7b83d2d0fde64153559ce3e31085ede/threedi_api_client/openapi/api_client.py#L76

**Comment by caspervdw** [comment]
> What's your opinion about using one of the retry policies as default?

That's a good idea, I defaulted to `3`, which is what `urllib3` does too.

Also I changed the API a bit, it is now a `retries` parameter instead of `rest_config["retries"]`. I see no other usage of `rest_config`currently and we can easily add more parameters if necessary.

**Pull request merged**